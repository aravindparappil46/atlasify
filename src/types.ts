import type { Notification, Reason } from './utils/api/typesGitHub';
import type { AuthMethod, PlatformType } from './utils/auth/types';

declare const __brand: unique symbol;

type Brand<B> = { [__brand]: B };

export interface AuthState {
  accounts: Account[];
}

export type Branded<T, B> = T & Brand<B>;

export type Username = Branded<string, 'Username'>;

export type Token = Branded<string, 'Token'>;

export type Organization = Branded<string, 'Organization'>;

export type Link = Branded<string, 'WebUrl'>;

export type Status = 'loading' | 'success' | 'error';

export interface Account {
  method: AuthMethod;
  platform: PlatformType;
  version?: string;
  token: Token;
  user: AtlasifyUser | null;
}

export type SettingsValue =
  | boolean
  | number
  | GroupBy
  | OpenPreference
  | Reason[]
  | Theme;

export type SettingsState = AppearanceSettingsState &
  NotificationSettingsState &
  SystemSettingsState &
  FilterSettingsState;

interface AppearanceSettingsState {
  theme: Theme;
  zoomPercentage: number;
  detailedNotifications: boolean;
  showAccountHeader: boolean;
  showPills: boolean;
  showNumber: boolean;
}

interface NotificationSettingsState {
  groupBy: GroupBy;
  participating: boolean;
  markAsDoneOnOpen: boolean;
  markAsDoneOnUnsubscribe: boolean;
  delayNotificationState: boolean;
}

interface SystemSettingsState {
  openLinks: OpenPreference;
  keyboardShortcut: boolean;
  showNotificationsCountInTray: boolean;
  showNotifications: boolean;
  useAlternateIdleIcon: boolean;
  playSound: boolean;
  openAtStartup: boolean;
}

interface FilterSettingsState {
  hideBots: boolean;
  filterReasons: Reason[];
}

export interface AtlasifyState {
  auth?: AuthState;
  settings?: SettingsState;
}

export enum Theme {
  SYSTEM = 'SYSTEM',
  LIGHT = 'LIGHT',
  DARK = 'DARK',
}

export enum OpenPreference {
  FOREGROUND = 'FOREGROUND',
  BACKGROUND = 'FOREGROUND',
}

export enum GroupBy {
  REPOSITORY = 'REPOSITORY',
  DATE = 'DATE',
}

export type RadioGroupItem = {
  label: string;
  value: string;
};

export interface AccountNotifications {
  account: Account;
  notifications: Notification[];
  error: AtlasifyError | null;
}

export interface AtlasifyUser {
  login: string;
  name: string | null;
  avatar: Link | null;
  id: string;
}

export interface AtlasifyError {
  title: string;
  descriptions: string[];
  emojis: string[];
}

export type ErrorType =
  | 'BAD_CREDENTIALS'
  | 'MISSING_SCOPES'
  | 'NETWORK'
  | 'RATE_LIMITED'
  | 'UNKNOWN';

export interface FormattedReason {
  title: string;
  description: string;
}

// TODO - Remove this
export enum IconColor {
  GRAY = 'text-gray-500 dark:text-gray-300',
  GREEN = 'text-green-500',
  PURPLE = 'text-purple-500',
  RED = 'text-red-500',
  YELLOW = 'text-yellow-500 dark:text-yellow-300',
  WHITE = 'text-white',
}

export enum Opacity {
  READ = 'opacity-50',
  LOW = 'opacity-70',
  MEDIUM = 'opacity-80',
  HIGH = 'opacity-90',
}

export enum Size {
  XSMALL = 12,
  SMALL = 14,
  MEDIUM = 16,
  LARGE = 18,
  XLARGE = 20,
}
