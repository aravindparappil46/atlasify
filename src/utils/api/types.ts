import type { Account, Link } from '../../types';

export interface AtlassianProduct {
  name: string;
  icon: React.ComponentType;
}

// Note: This is not in the official GitHub API. We add this to make notification interactions easier.
export interface GitifyNotification {
  account: Account;
}

export interface MyUserDetails {
  me: {
    user: {
      accountId: string;
      name: string;
      picture: Link;
    };
  };
}

export interface MyNotifications {
  notifications: {
    unseenNotificationCount: number;
    notificationFeed: {
      nodes: AtlassianNotification[];
    };
  };
}

export type Category = 'direct' | 'watching';

export type ReadState = 'unread' | 'read';

export type Product =
  | 'bitbucket'
  | 'confluence'
  | 'compass'
  | 'jira'
  | 'unknown';

export interface AtlassianNotification {
  groupId: string;
  headNotification: {
    notificationId: string;
    timestamp: string;
    readState: string;
    category: string;
    content: {
      type: string;
      message: string;
      url: Link;
      path: [
        {
          title: string;
          url: Link;
          iconUrl: Link | null;
        },
      ];
      entity: {
        title: string;
        iconUrl: Link;
        url: Link;
      };
      actor: {
        displayName: string;
        avatarURL: Link;
      };
    };
    analyticsAttributes: [
      {
        key: string;
        value: string;
      },
    ];
  };
}

export type UserDetails = User & UserProfile;

export interface UserProfile {
  name: string;
  company: string;
  blog: string;
  location: string;
  email: string;
  hireable: string;
  bio: string;
  twitter_username: string;
  public_repos: number;
  public_gists: number;
  followers: number;
  following: number;
  created_at: string;
  updated_at: string;
  private_gists: number;
  total_private_repos: number;
  owned_private_repos: number;
  disk_usage: number;
  collaborators: number;
  two_factor_authentication: boolean;
  plan: Plan;
}

export interface Plan {
  name: string;
  space: number;
  private_repos: number;
  collaborators: number;
}

export interface User {
  login: string;
  id: number;
  node_id: string;
  avatar_url: Link;
  gravatar_url: Link;
  url: Link;
  html_url: Link;
  followers_url: Link;
  following_url: Link;
  gists_url: Link;
  starred_url: Link;
  subscriptions_url: Link;
  organizations_url: Link;
  repos_url: Link;
  events_url: Link;
  received_events_url: Link;
  type: UserType;
  site_admin: boolean;
}

export type UserType =
  | 'Bot'
  | 'EnterpriseUserAccount'
  | 'Mannequin'
  | 'Organization'
  | 'User';

export interface SubjectUser {
  login: string;
  html_url: Link;
  avatar_url: Link;
  type: UserType;
}

export interface Repository {
  id: number;
  node_id: string;
  name: string;
  full_name: string;
  private: boolean;
  owner: Owner;
  html_url: Link;
  description: string;
  fork: boolean;
  url: Link;
  forks_url: Link;
  keys_url: Link;
  collaborators_url: Link;
  teams_url: Link;
  hooks_url: Link;
  issue_events_url: Link;
  events_url: Link;
  assignees_url: Link;
  branches_url: Link;
  tags_url: Link;
  blobs_url: Link;
  git_tags_url: Link;
  git_refs_url: Link;
  trees_url: Link;
  statuses_url: Link;
  languages_url: Link;
  stargazers_url: Link;
  contributors_url: Link;
  subscribers_url: Link;
  subscription_url: Link;
  commits_url: Link;
  git_commits_url: Link;
  comments_url: Link;
  issue_comment_url: Link;
  contents_url: Link;
  compare_url: Link;
  merges_url: Link;
  archive_url: Link;
  downloads_url: Link;
  issues_url: Link;
  pulls_url: Link;
  milestones_url: Link;
  notifications_url: Link;
  labels_url: Link;
  releases_url: Link;
  deployments_url: Link;
}

export interface Owner {
  login: string;
  id: number;
  node_id: string;
  avatar_url: Link;
  gravatar_id: string;
  url: Link;
  html_url: Link;
  followers_url: Link;
  following_url: Link;
  gists_url: Link;
  starred_url: Link;
  subscriptions_url: Link;
  organizations_url: Link;
  repos_url: Link;
  events_url: Link;
  received_events_url: Link;
  type: string;
  site_admin: boolean;
}

export type Subject = GitHubSubject & GitifySubject;

interface GitHubSubject {
  title: string;
  url: Link | null;
  latest_comment_url: Link | null;
}

// This is not in the GitHub API, but we add it to the type to make it easier to work with
export interface GitifySubject {
  number?: number;
  user?: SubjectUser;
  linkedIssues?: string[];
  comments?: number;
  labels?: string[];
}

export interface GraphQLSearch<T> {
  data: {
    search: {
      nodes: T[];
    };
  };
}

export interface GraphQLResponse<T> {
  data: T;
}

export interface GitHubRESTError {
  message: string;
  documentation_url: Link;
}
