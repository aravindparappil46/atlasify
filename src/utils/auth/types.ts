import type { Token, Username } from '../../types';

export type AuthMethod = 'API Token';

export type PlatformType = 'Atlassian Cloud';

export interface LoginAPITokenOptions {
  username: Username;
  token: Token;
}
