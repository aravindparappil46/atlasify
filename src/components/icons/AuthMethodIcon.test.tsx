import { render } from '@testing-library/react';
import { Size } from '../../types';
import { AuthMethodIcon, type IAuthMethodIcon } from './AuthMethodIcon';

describe('components/icons/AuthMethodIcon.tsx', () => {
  it('should render API Token icon', () => {
    const props: IAuthMethodIcon = {
      type: 'API Token',
      size: Size.MEDIUM,
    };
    const tree = render(<AuthMethodIcon {...props} />);
    expect(tree).toMatchSnapshot();
  });
});
