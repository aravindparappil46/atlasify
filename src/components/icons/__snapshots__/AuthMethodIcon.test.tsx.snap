// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/icons/AuthMethodIcon.tsx should render API Token icon 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="mr-1"
        title="API Token"
      >
        <svg
          aria-hidden="true"
          class="octicon octicon-key"
          fill="currentColor"
          focusable="false"
          height="16"
          style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
          viewBox="0 0 16 16"
          width="16"
        >
          <path
            d="M10.5 0a5.499 5.499 0 1 1-1.288 10.848l-.932.932a.749.749 0 0 1-.53.22H7v.75a.749.749 0 0 1-.22.53l-.5.5a.749.749 0 0 1-.53.22H5v.75a.749.749 0 0 1-.22.53l-.5.5a.749.749 0 0 1-.53.22h-2A1.75 1.75 0 0 1 0 14.25v-2c0-.199.079-.389.22-.53l4.932-4.932A5.5 5.5 0 0 1 10.5 0Zm-4 5.5c-.001.431.069.86.205 1.269a.75.75 0 0 1-.181.768L1.5 12.56v1.69c0 .138.112.25.25.25h1.69l.06-.06v-1.19a.75.75 0 0 1 .75-.75h1.19l.06-.06v-1.19a.75.75 0 0 1 .75-.75h1.19l1.023-1.025a.75.75 0 0 1 .768-.18A4 4 0 1 0 6.5 5.5ZM11 6a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"
          />
        </svg>
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="mr-1"
      title="API Token"
    >
      <svg
        aria-hidden="true"
        class="octicon octicon-key"
        fill="currentColor"
        focusable="false"
        height="16"
        style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
        viewBox="0 0 16 16"
        width="16"
      >
        <path
          d="M10.5 0a5.499 5.499 0 1 1-1.288 10.848l-.932.932a.749.749 0 0 1-.53.22H7v.75a.749.749 0 0 1-.22.53l-.5.5a.749.749 0 0 1-.53.22H5v.75a.749.749 0 0 1-.22.53l-.5.5a.749.749 0 0 1-.53.22h-2A1.75 1.75 0 0 1 0 14.25v-2c0-.199.079-.389.22-.53l4.932-4.932A5.5 5.5 0 0 1 10.5 0Zm-4 5.5c-.001.431.069.86.205 1.269a.75.75 0 0 1-.181.768L1.5 12.56v1.69c0 .138.112.25.25.25h1.69l.06-.06v-1.19a.75.75 0 0 1 .75-.75h1.19l.06-.06v-1.19a.75.75 0 0 1 .75-.75h1.19l1.023-1.025a.75.75 0 0 1 .768-.18A4 4 0 1 0 6.5 5.5ZM11 6a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"
        />
      </svg>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
